pkg/apis/resource/validation/validation.go

```diff
@@ -127,19 +127,25 @@ func ValidateClaimStatusUpdate(resourceClaim, oldClaim *resource.ResourceClaim)
 	// The name might not be set yet.
 	if resourceClaim.Status.DriverName != "" {
 		allErrs = append(allErrs, validateResourceDriverName(resourceClaim.Status.DriverName, fldPath.Child("driverName"))...)
-	} else if resourceClaim.Status.Allocation != nil {
-		allErrs = append(allErrs, field.Required(fldPath.Child("driverName"), "must be specified when `allocation` is set"))
+	} else if len(resourceClaim.Status.Allocation) != 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("driverName"), "must be specified when `allocation` is populated"))
 	}
 
 	allErrs = append(allErrs, validateAllocationResult(resourceClaim.Status.Allocation, fldPath.Child("allocation"))...)
 	allErrs = append(allErrs, validateResourceClaimConsumers(resourceClaim.Status.ReservedFor, resource.ResourceClaimReservedForMaxSize, fldPath.Child("reservedFor"))...)
 
 	// Now check for invariants that must be valid for a ResourceClaim.
 	if len(resourceClaim.Status.ReservedFor) > 0 {
-		if resourceClaim.Status.Allocation == nil {
-			allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "may not be specified when `allocated` is not set"))
+		if len(resourceClaim.Status.Allocation) == 0 {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "may not be specified when `allocation` is empty"))
 		} else {
-			if !resourceClaim.Status.Allocation.Shareable && len(resourceClaim.Status.ReservedFor) > 1 {
+			// TODO: @pohly what should we do here now that there may be more
+			// than one (possibly conflicting) 'Sharable' settings in each
+			// 'AllocationResult'?  Should we verify that all 'Sharable' values
+			// are the same? Or pull the 'Sharable' field up to the
+			// 'ResourceClaimStatus' struct instead of embeddng it in each
+			// 'AllocationResult'?
+			if !resourceClaim.Status.Allocation[0].Shareable && len(resourceClaim.Status.ReservedFor) > 1 {
 				allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "may not be reserved more than once"))
 			}
```

Comment:
@pohly , @bart0sh this is the biggest question I have about this change.

Should we reconsider making `Allocation` an '[]AllocationResult` and instead do something more similar to what Ed had suggested at one point.

I.e. transform the embedded `ResourceHandle` into a `[]ResourceHandle`, where `ResourceHandle` is defined as:
```
type ResourceHandle struct {
    KubeletPluginName string
    Handle string
}
```
In this way, only the handle becomes Kubelet plugin specific, rather than all fields of the AllocationResult.

At one point I argued that it made sense to make `Sharable` and `AvailableOnNodes` parameterized on the `KubeletPluginName` as well (which is why we turned the whole `AllocationResult` into a slice), but after working through this PR I'm not convinced it makes sense anymore and the change above is probably the better change.

Reply:
WIP implementation of this alternative idea:
https://github.com/pohly/kubernetes/compare/dra-v1alpha2...klueska:kubernetes:extend-resourceclaimstatus-2

Reply:
> transform the embedded ResourceHandle into a []ResourceHandle

I like this better. It clearly shows that there could be more than one allocation result, one per driver.

However, we may want to look again and your reasoning when I proposed similar change. I can't find those comments for some reason.

Reply:
@pohly What do you think? My main issue with this is that I (as before) don't really like the name `ResourceHandle`. Could we rename it something like `NodePayload` or something like that?

Reply:
In a way, the current `Shareable` is a placeholder: the idea was to extend this later with more specific usage constraints, like "accessible on any node, but only usable on one node at a time, by an arbitrary number of pods".

The way the API is designed now, the "meta DRA driver controller" doesn't need to do anything with this information. It can simply get it from the underlying controllers and store it. If we only have one `Shareable` or (in the future) one set of constraints, then the meta controller would have to be aware of all that logic and merge it.  This is the exact same problem that Kubernetes has now.

Perhaps moving it into the controller is the right choice because perhaps it has additional knowledge that Kubernetes doesn't have. So I am tentatively okay with going in that direction.

Regarding naming: `ResourceHandle` was chosen when it was a single string. Perhaps `ResourceAttributes`, similar to [`VolumeAttributes`](https://github.com/kubernetes/kubernetes/blob/713ded736824d669d9fae9ac4a29b1ec18d66446/staging/src/k8s.io/api/core/v1/types.go#L1800)?

I'm not aware of another API field that uses "payload".

Reply:
I am actually fine with the type being called `ResourceHandle`, it is more about the name of the field in the `AllocationResult`.

So rather than this:
```
type AllocationResult struct {
    ResourceHandles []ResourceHandle 
    AvailableOnNodes *core.NodeSelector
    Shareable bool
}

type ResourceHandle struct {
    KubeletPluginName string
    Data string
}
```
Having something like this:
```
type AllocationResult struct {
    NodePluginData []ResourceHandle 
    AvailableOnNodes *core.NodeSelector
    Shareable bool
}

type ResourceHandle struct {
    PluginName string
    Data string
}
```
i.e. something that indicates this data is destined for the kubelet plugin.

Reply:
The data is not just for the node. When it comes to deallocation, a controller also might need the information that it stored earlier during allocation. That makes `NodePluginData` a bad name.

Reply:
That's fair -- what about something as simple as `Context`, i.e.
```
type AllocationResult struct {
    Context []ResourceHandle 
    AvailableOnNodes *core.NodeSelector
    Shareable bool
}

type ResourceHandle struct {
    KubeletPluginName string
    Data string
}
```

Reply:
OK; I will go with that then and update everything as appropriate (implementation PR and KEP update PR)

Reply:
I believe it is okay to first do the implementation PR, get that reviewed, and then update the KEP.

Reply:
Yep, that's my plan...

Reply:
The latest code is now in line with what we've discussed here. Resolving this comment.